source/change-streams/tests/README.rst

```diff
@@ -0,0 +1,156 @@
+  - ``result``: Document with ONE of the following fields:
+
+    - ``error``: Describes an error received during the test
+    - ``success``: An array of documents expected to be received from the changeStream
+
+Spec Test Match Function
+========================
+
+The definition of MATCH or MATCHES in the Spec Test Runner is as follows:
+
+- MATCH takes two values, ``expected`` and ``actual``
+- Notation is "Assert [actual] MATCHES [expected]
+- Assertion passes if ``expected`` is a subset of ``actual``, with the values ``42`` and ``"42"`` acting as placeholders for "any value"
+
+Pseudocode Implementation of ``actual`` MATCHES ``expected``:
+
+- If ``expected`` is ``"42"`` or ``42``
+
+  - Assert that ``actual`` exists (is not ``null`` or ``undefined``)
+
+- Else
+
+  - Assert that ``actual`` is of the same JSON type as ``expected``
+  - If ``expected`` is a ``number``, ``string``, ``boolean`` or ``null``
+
+    - Assert that ``expected`` equals ``actual``
+
+  - Else if ``expected`` is an ``array``
+
+    - for every ``idx``/``value`` in ``expected``
+
+      - Assert that ``value`` MATCHES ``actual[idx]``
+
+  - Else
+
+    - for every ``key``/``value`` in ``expected``
+
+      - Assert that ``value`` MATCHES ``actual``
+
+Spec Test Runner
+================
+
+Before running the tests
+
+- Create a MongoDB server topology.
+- Create a MongoClient ``globalClient``, and connect to the server
+
+For each YAML file, for each element in ``tests``:
+
+- If ``topology`` does not include the topology of the server instance(s), skip this test.
+- Use ``globalClient`` to
+
+  - Drop the database ``database_name``
+  - Create the database ``database_name`` and the collection ``database_name.collection_name``
+
+- Create a new MongoClient ``client``, with APM enabled
+- Using ``client``, create a changeStream ``changeStream`` against the specified ``target``. Use ``changeStreamPipeline`` and ``changeStreamOptions`` if they are non-empty
+- Using ``globalClient``, run every operation in ``operations`` in serial against the server
+- Wait until either
+
+  - An error occurres
+  - All operations have been successful AND the changeStream has received as many changes as there are ``result.success``
+
+- Close ``changeStream``
+- If there was an error
+
+  - Assert that an error was expected for the test.
+  - Assert that the error MATCHES ``results.error``
+
+- Else
+
+  - Assert that no error was expected for the test
+  - Assert that the changes received from ``changeStream`` MATCH the results in ``results.success``
+
+- If there are any ``expectations``
+
+  - For each APM ``commandStarted`` events on ``client``
+
+    - Filter out any ``ismaster`` events
+
+  - For each (``expected``, ``idx``) in ``expectations``
+
+    - Assert that ``actual[idx]`` MATCHES ``expected``
+
+- Close the MongoClient ``client``
+
+After running all tests
+
+- Close the MongoClient ``globalClient``
+- Shut down the server topology
```

Comment:
This reads as if drivers should terminate the entire cluster (e.g. via Mongo Orchestration). Is that really necessary? Wouldn't it be sufficient to drop the relevant databases?

Reply:
I only state this b/c I said `Create a MongoDB server topology.` before running the tests. I could remove both references and instead say to drop the databases after the tests.